shared/js/dialer/keypad.js

```diff
@@ -504,6 +517,105 @@ var KeypadManager = {
     }
   },
 
+  /**
+   * Returns true if the number is a speed dial code as described in
+   * 3GPP TS 22.030 6.6.4. Speed dial codes are in the N(N)(N)# format.
+   */
+  _isSpeedDialNumber: function(number) {
+    return !!number.match(/^(0|[1-9][0-9]{0,2})\#$/);
+  },
+
+  /**
+   * Returns the telephony number corresponding to the specified index. Speed
+   * dial numbers are retrieved from the SIM contacts list and not from the
+   * regular contacts.
+   *
+   * @param {Integer} cardIndex The SIM card index.
+   * @param {Integer} index The index of the speed dial number.
+   * @returns {Promise} A promise that resolves to the corresponding number.
+   */
+  _getSpeedDialNumber: function(cardIndex, index) {
+    var self = this;
+
+    index--; // Speed dial indexes seem to be 1-based
+
+    return this._populateSimContactsLists(cardIndex).then(function() {
+      if (index < self._simContactsList[cardIndex].length) {
+        return self._simContactsList[cardIndex][index].number;
+      } else {
+        return Promise.reject();
+      }
+    });
+  },
+
+  /**
+   * Populates the SIM contacts list for the specified SIM card.
+   *
+   * @param {Integer} cardIndex The SIM card index.
+   * @returns {Promise} A promise that is resolved once the list is populated.
+   */
+  _populateSimContactsLists: function(cardIndex) {
+    if ((this._simContactsList.length > cardIndex) &&
+        this._simContactsList[cardIndex]) {
+      return Promise.resolve();
+    }
+
+    var self = this;
+    var canceled = false;
+
+    return new Promise(function(resolve, reject) {
+      LazyLoader.load(['/shared/style/confirm.css',
+                       '/shared/js/confirm.js',
+                       document.getElementById('confirmation-message')],
+        function() {
+          var iccId = navigator.mozIccManager.iccIds[cardIndex];
+          var icc = navigator.mozIccManager.getIccById(iccId);
+          var req = icc.readContacts('adn');
+
+          req.onsuccess = function(event) {
+            var adnContacts = event.target.result;
+            var numbers = new Array(adnContacts.length);
+
+            for (var i = 0; i < adnContacts.length; i++) {
+              numbers[i] = {
+                id: adnContacts[i].id,
+                number: adnContacts[i].tel[0].value,
+              };
+            }
+
+            numbers.sort(function(a, b) {
+              if (a.id.length == b.id.length) {
+                return (a.id > b.id) ? 1 : 0;
+              } else {
+                return (a.id.length > b.id.length) ? 1 : 0;
+              }
+            });
+
+            self._simContactsList[cardIndex] = numbers;
+
+            if (!canceled) {
+              ConfirmDialog.hide();
+            }
+
+            resolve();
+          };
+          req.onerror = function() {
+            reject();
```

Comment:
This could use a ```console.error()``` to explain what happened. I think the ```onerror()``` callback returns an actual error, too.